lib/src/summation/counter.dart

```diff
@@ -0,0 +1,157 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// counter.dart
+// A flexible counter implementation.
+//
+// 2024 August 26
+// Author: Max Korbel <max.korbel@intel.com>
+
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+import 'package:rohd_hcl/src/summation/summation_base.dart';
+
+/// Keeps a count of the running sum of any number of sources with optional
+/// configuration for widths, saturation behavior, and restarting.
+class Counter extends SummationBase {
+  /// The output value of the counter.
+  Logic get count => output('count');
+
+  /// Creates a counter that increments according to the provided [interfaces].
+  ///
+  /// The [width] can be either explicitly provided or inferred from other
+  /// values such as a [maxValue], [minValue], or [resetValue] that contain
+  /// width information (e.g. a [LogicValue]), or by making it large enough to
+  /// fit [maxValue], or by inspecting widths of [interfaces]. There must be
+  /// enough information provided to determine the [width].
+  ///
+  /// If no [maxValue] is provided, one will be inferred by the maximum that can
+  /// fit inside of the [width].
+  ///
+  /// The [restart] input can be used to restart the counter to a new value, but
+  /// also continue to increment in that same cycle. This is distinct from
+  /// [reset] which will reset the counter, holding the [count] at [resetValue].
+  ///
+  /// If [saturates] is `true`, then it will saturate at the [maxValue] and
+  /// [minValue]. If `false`, will wrap around (overflow/underflow) at the
+  /// [maxValue] and [minValue].  The [equalsMax], [equalsMin], [overflowed],
+  /// and [underflowed] outputs can be used to determine if the sum is at the
+  /// maximum, minimum, (would have) overflowed, or  (would have) underflowed,
+  /// respectively.
+  Counter(
+    super.interfaces, {
+    required Logic clk,
+    required Logic reset,
+    Logic? restart,
+    dynamic resetValue = 0,
+    super.maxValue,
+    super.minValue = 0,
+    super.width,
+    super.saturates,
+    super.name = 'counter',
+  }) : super(initialValue: resetValue) {
+    clk = addInput('clk', clk);
+    reset = addInput('reset', reset);
+
+    if (restart != null) {
+      restart = addInput('restart', restart);
+    }
+
+    addOutput('count', width: width);
+
+    final sum = Sum(
+      interfaces,
+      initialValue:
+          restart != null ? mux(restart, initialValueLogic, count) : count,
+      maxValue: maxValueLogic,
+      minValue: minValueLogic,
+      width: width,
+      saturates: saturates,
+    );
+
+    count <=
+        flop(
+          clk,
+          sum.sum,
+          reset: reset,
+          resetValue: initialValueLogic,
+        );
+
+    // need to flop these since value is flopped
+    overflowed <= flop(clk, sum.overflowed, reset: reset);
```

Comment:
If we had a different word for this, we might collapse under/overflow into a single output since the Counter runs in one direction.

Not sure now about equalsMax/Min:  if we have an incrementing counter, do we see equalsMin on reset? Or is this really and 'equalsLimit' and we could collapse these two into one output.

Reply:
We need both because we could have both incrementing and decrementing inputs, with variable amounts and enables, so it's possible for the same counter to experience min and max both